src/y0/algorithm/identify/cg.py

```diff
@@ -0,0 +1,318 @@
+                    if node_at_intervention2 in new_event:
+                        # TODO needs test case
+                        new_event[node_at_intervention1] = new_event[node_at_intervention2]
+                        new_event.pop(node_at_intervention2, None)
+    rv_graph = cf_graph.subgraph(cf_graph.ancestors_inclusive(new_event))
+    # rv_graph = rv_graph.remove_nodes_from(
+    #    node for node in rv_graph.nodes() if is_self_intervened(node)
+    # )
+    # rv_graph = remove_redundant_interventions(rv_graph)
+    return rv_graph, new_event
+
+
+def remove_redundant_interventions(graph: NxMixedGraph) -> NxMixedGraph:
+    # TODO this function is not used. delete?
+    # TODO write documentation on the goal of this function
+    relabel: Dict[CounterfactualVariable, Variable] = {}
+    for node in graph.topological_sort():
+        if not isinstance(node, CounterfactualVariable):
+            continue
+        ancestor_bases = [ancestor.get_base() for ancestor in graph.ancestors_inclusive(node)]
+        # Things to check:
+        #  1. does not have 0 interventions, 1 intervention, or more than one?
+        #  a. is node already in relabeling?
+        #  2b. if so, is it a regular variable or a counterfactual variable?
+        if 1 == len(node.interventions):
+            intervention = node.interventions[0]
+            raise NotImplementedError
+        else:
+            for intervention in node.interventions:
+                if intervention.get_base() in ancestor_bases:
+                    continue
+                # This next part does not make sense, the same node would get
+                # overwritten in the relabel dictionary for each intervention
+                if 1 == len(node.interventions) or (
+                    node in relabel and 1 == len(relabel[node].interventions)
+                ):
+                    relabel[node] = Variable(name=node.name)
+                elif node not in relabel:
+                    relabel[node] = CounterfactualVariable(
+                        name=node.name,
+                        interventions=tuple(sorted(set(node.interventions) - {intervention})),
+                    )
+                else:
+                    relabel[node] = CounterfactualVariable(
+                        name=node.name,
+                        interventions=tuple(
+                            sorted(set(relabel[node].interventions) - {intervention})
+                        ),
+                    )
+    return graph.relabel_nodes(relabel)
+
+
+def make_parallel_worlds_graph(graph: NxMixedGraph, worlds: Worlds) -> NxMixedGraph:
+    """Make a parallel worlds graph.
+
+    :param graph: A normal graph
+    :param worlds: A set of sets of treatments
+    :returns: A combine parallel world graph
+    """
+    world_graphs: List[NxMixedGraph] = [graph.intervene(world) for world in worlds]
+    return combine_worlds(graph, world_graphs, worlds)
+
+
+def _stitch_1(graph: NxMixedGraph, worlds: Worlds):
```

Comment:
@djinnome I refactored these 4 stitching operations out of the main `combine_worlds()` function. Can you please give some high level narration for each of them as docstrings?